### PR TITLE
psh: add ls command test

### PR DESCRIPTION
**Pull request by damianloew**
JIRA: PD-77

## Description

- psh: add ls command test, add get_commands function to psh module
- psh: add ls command test for rootfs targets, move esc_codes dictionary to psh module
- psh: simplify ls command tests

## Motivation and Context

It's required for psh commands tests
## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## How Has This Been Tested?

- [ ] Already covered by automatic testing.
- [ ] New test added: (add PR link here).
- [x] Tested by hand on: (ia32-generic - qemu).

## Checklist:

- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [x] All new and existing linter checks and tests passed.
- [x] My changes generate no new compilation warnings for any of the targets.

## Special treatment

- [ ] This PR needs additional PRs to work (list the PRs, preferably in merge-order).
- [ ] I will merge this PR by myself when appropriate.

Related to the following issues:
https://github.com/phoenix-rtos/phoenix-rtos-project/issues/190
https://github.com/phoenix-rtos/phoenix-rtos-project/issues/194

### CODE REVIEW

File: psh/test-ls.py
Comment (nalajcie) [inline psh/test-ls.py:316]: this is invalid as it assumes there are ONLY psh commands in the result. When You would build the full project there will be for example `busybox` binary between `bind` and `cat`.

Also - I would add param to disable coloring in psh `ls` (the same way it's done in buxybox ls) to make the parsing in tests easier.

Comment (damianloew) [reply psh/test-ls.py]: That's right, now it would works well and check if all commands from PSH_COMMANDS list are visible (even if between them are other files)

So do You think it's better to disable coloring when testing and don't test this functionality? Tests, which parse colored output are already written, I would leave leave it as it is. However, I agree param like that in ls would be a good idea.

**Review by xvuko — COMMENTED** [review]
I do not think this is correct approach to functional testing.

This test is re-implementing psh `ls` command in a way that has exactly same binary compatible output for single fixed list of arguments as current C version. This is nothing more than disguised and over-complicated regression testing. Using this test will make improving `ls` command a tedious an painful experience.

I would expect this to look something like:
1. Prepare test directory
2. Execute tested command grabbing all output without parsing it
3. strip escape sequences
4. normalize white-space (white-space in files would need to be addressed somehow)
5. parse files from normalized output
6. compare expected list to parsed one

Would be good to involve some other people in this discussion.

**Comment by damianloew** [comment]
> I do not think this is correct approach to functional testing.
> 
> This test is re-implementing psh `ls` command in a way that has exactly same binary compatible output for single fixed list of arguments as current C version. This is nothing more than disguised and over-complicated regression testing. Using this test will make improving `ls` command a tedious an painful experience.
> 
> I would expect this to look something like:
> 
> 1. Prepare test directory
> 2. Execute tested command grabbing all output without parsing it
> 3. strip escape sequences
> 4. normalize white-space (white-space in files would need to be addressed somehow)
> 5. parse files from normalized output
> 6. compare expected list to parsed one
> 
> Would be good to involve some other people in this discussion.



> I do not think this is correct approach to functional testing.
> 
> This test is re-implementing psh `ls` command in a way that has exactly same binary compatible output for single fixed list of arguments as current C version. This is nothing more than disguised and over-complicated regression testing. Using this test will make improving `ls` command a tedious an painful experience.
> 
> I would expect this to look something like:
> 
> 1. Prepare test directory
> 2. Execute tested command grabbing all output without parsing it
> 3. strip escape sequences
> 4. normalize white-space (white-space in files would need to be addressed somehow)
> 5. parse files from normalized output
> 6. compare expected list to parsed one
> 
> Would be good to involve some other people in this discussion.

Thanks for the review. This test was written a long time ago, I didn't want to spend much time on changing it and merge it to start regression testing earlier. But you are right, it needs to be more flexible. I think regular expressions and its support in the pexpect module can be sufficient here. I realized that maybe checking colors and exact esc sequences isn't crucial and now there can be various white spaces and esc sequences between printed files. Please check the latest commit, it is my proposal, I hope it solves this problem.

File: psh/test-ls.py
Comment (damianloew) [reply psh/test-ls.py]: The current version of test supports both solutions.

File: psh/test-ls.py
Comment (damianloew) [reply psh/test-ls.py]: I've moved common parts of code to an additional function in next commit.

File: psh/test-ls.py
Comment (xvuko) [inline psh/test-ls.py:110]: There can be two spaces before date if listening contains files with two digit days

File: psh/test-ls.py
Comment (xvuko) [inline psh/test-ls.py:132]: This is not checking files with different sizes. If for some reason we have no way to properly test this now then it should be clearly stated in comment.

File: psh/test-ls.py
Comment (xvuko) [inline psh/test-ls.py:128]: This is a noop for files used in harness as dir would be first after sorting anyway. This could create invalid order in other cases:

`('a', 'b', 'dir')` would become `('dir', 'b', 'a')`


Key parameter can be used to modify sorting rules:
```py
size_sorted_files = list(sorted(files, key=lambda x: (0, x) if is_dir(x) else (1, x)))
```
sorted function in this example will compare first member of tuple first so directories are given "higher priority" here.

File: psh/test-ls.py
Comment (xvuko) [inline psh/test-ls.py:106]: `( )`

Groups around single characters are not needed.

**Comment by xvuko** [comment]
> Please check the latest commit, it is my proposal, I hope it solves this problem.

Yes with more flexible regex expressions it is ok now.

File: psh/test-ls-rootfs.py
Comment (damianloew) [reply psh/test-ls-rootfs.py:49]: You are right, I've used newly created date command for asserting it in last push (temporarily this test is skipped, so it can wait for date implementation merge).

File: psh/test-ls.py
Comment (damianloew) [reply psh/test-ls.py:59]: I've added mentioned cases. There were some problem with printing very long file names, so I've created issue and fix proposal. `assert_extralong(p)` works with it.

File: psh/test-ls.py
Comment (damianloew) [reply psh/test-ls.py:111]: I'm sorry, my mistake

File: psh/test-ls.py
Comment (damianloew) [reply psh/test-ls.py:106]: I thought it is more readable, but maybe You are right and it's better without parentheses.

File: psh/test-ls.py
Comment (damianloew) [reply psh/test-ls.py]: You are right, corrected.

File: psh/test-ls.py
Comment (xvuko) [inline psh/test-ls.py:130]: ''.join(arg_help_pattern) is equal to arg_help_pattern

File: psh/test-ls.py
Comment (xvuko) [inline psh/test-ls.py:140]: "." is not expected on `ls -l`

File: psh/test-ls-rootfs.py
Comment (damianloew) [reply psh/test-ls-rootfs.py:51]: I think about removing the second part of msg statement, because called command is printed in reader buffer, like here:
Screenshot from 2022-03-22 14-12-12

Comment (xvuko) [reply psh/test-ls-rootfs.py:51]: Most important thing to include in output from tested device. Everything else can be deduced from test source code. Its nice to have send input and expected output. Additional message should not be needed in most cases.

File: psh/test-ls-rootfs.py
Comment (damianloew) [reply psh/test-ls-rootfs.py]: Applied, thanks for the advice.

File: psh/test-ls-rootfs.py
Comment (damianloew) [reply psh/test-ls-rootfs.py]: I've corrected and enabled this test case (https://github.com/phoenix-rtos/phoenix-rtos-project/issues/194 seems to be fixed)

File: psh/test-ls-rootfs.py
Comment (damianloew) [reply psh/test-ls-rootfs.py]: I would like to leave this case, because there may occur some problem (like in https://github.com/phoenix-rtos/phoenix-rtos-project/issues/194). But I didn't have idea how to do it better than avoiding `exit` and `history` for now.

File: psh/test-ls.py
Comment (damianloew) [reply psh/test-ls.py]: Control code pattern applied

File: psh/test-ls.py
Comment (damianloew) [reply psh/test-ls.py]: I've done something like that. Please let me know if it's ok.
Screenshot from 2022-03-23 17-09-00

File: psh/test_ls.py
Comment (xvuko) [inline psh/test_ls.py:19]: (\s+)+ is an example of Evil Regex: https://en.wikipedia.org/wiki/ReDoS#Exponential_backtracking

**Review by xvuko — COMMENTED** [review]
Please do not add unrelated file renames to reviews.

File: psh/test_ls_rootfs.py
Comment (damianloew) [reply psh/test_ls_rootfs.py:82]: It's not possible to send lines longer than window size (80). That's why I've done it like this. Thanks, `[^\r\n]+` seems to be a better solution.

File: psh/test_ls_rootfs.py
Comment (damianloew) [reply psh/test_ls_rootfs.py]: Thanks, I've used both options in last commit.

**Comment by damianloew** [comment]
> Please do not add unrelated file renames to reviews.

Sorry, I will create another PR for that then.

File: psh/test_ls_rootfs.py
Comment (xvuko) [reply psh/test_ls_rootfs.py]: I mean that r'([^\r\n]+(\r+\n)){2,}' would be more readable.

File: psh/test_ls_rootfs.py
Comment (xvuko) [reply psh/test_ls_rootfs.py]: `file{i}` in assert_cmd comment should be changed to `{file}`

File: psh/test_ls.py
Comment (xvuko) [reply psh/test_ls.py:107]:
```py
>>> re.match(r'(.){2}', 'ab').groups()
('b',)
```

**Review by xvuko — APPROVED** [review]
Added minor comments. Can be merged after addressing them.

File: psh/test_ls.py
Comment (damianloew) [reply psh/test_ls.py]: I'd like to leave it to improve readability. Thanks, I've verified that `(?P<permissions>([-r][-w][-x]){3})` is matched fine, so I've left it like that.

**Comment by damianloew** [comment]
> Added minor comments. Can be merged after addressing them.

Thanks, I've done it. Last three commits are squashed, ready for final approve and merge.

**Review by xvuko — APPROVED** [review]

**Pull request merged**
